Snowflake.Data.Tests/IntegrationTests/StructuredTypesIT.cs

```diff
@@ -0,0 +1,1471 @@
+                    Assert.IsTrue(reader.Read());
+
+                    // act
+                    var address = reader.GetObject<Address>(0);
+
+                    // assert
+                    Assert.AreEqual("San Mateo", address.city);
+                    Assert.AreEqual("CA", address.state);
+                    Assert.IsNull(address.zip);
+                }
+            }
+        }
+
+        [Test]
+        public void TestSelectNestedStructuredTypeObject()
+        {
+            // arrange
+            using (var connection = new SnowflakeDbConnection(ConnectionString))
+            {
+                connection.Open();
+                using (var command = connection.CreateCommand())
+                {
+                    EnableStructuredTypes(connection);
+                    var addressAsSFString =
+                        "OBJECT_CONSTRUCT('city','San Mateo', 'state', 'CA', 'zip', OBJECT_CONSTRUCT('prefix', '00', 'postfix', '11'))::OBJECT(city VARCHAR, state VARCHAR, zip OBJECT(prefix VARCHAR, postfix VARCHAR))";
+                    command.CommandText = $"SELECT {addressAsSFString}";
+                    var reader = (SnowflakeDbDataReader)command.ExecuteReader();
+                    Assert.IsTrue(reader.Read());
+
+                    // act
+                    var address = reader.GetObject<Address>(0);
+
+                    // assert
+                    Assert.AreEqual("San Mateo", address.city);
+                    Assert.AreEqual("CA", address.state);
+                    Assert.NotNull(address.zip);
+                    Assert.AreEqual("00", address.zip.prefix);
+                    Assert.AreEqual("11", address.zip.postfix);
+                }
+            }
+        }
+
+        [Test]
+        public void TestSelectObjectWithMap()
+        {
+            // arrange
+            using (var connection = new SnowflakeDbConnection(ConnectionString))
+            {
+                connection.Open();
+                using (var command = connection.CreateCommand())
+                {
+                    EnableStructuredTypes(connection);
+                    var objectWithMap = "OBJECT_CONSTRUCT('names', OBJECT_CONSTRUCT('Excellent', '6', 'Poor', '1'))::OBJECT(names MAP(VARCHAR,VARCHAR))";
+                    command.CommandText = $"SELECT {objectWithMap}";
+                    var reader = (SnowflakeDbDataReader)command.ExecuteReader();
+                    Assert.IsTrue(reader.Read());
+
+                    // act
+                    var grades = reader.GetObject<GradesWithMap>(0);
+
+                    // assert
+                    Assert.NotNull(grades);
+                    Assert.AreEqual(2, grades.Names.Count);
+                    Assert.AreEqual("6", grades.Names["Excellent"]);
+                    Assert.AreEqual("1", grades.Names["Poor"]);
+                }
+            }
+        }
+
+        [Test]
+        public void TestSelectArray()
```

Comment:
Does it work for all types? Especially binary and date/time/timestamps?

Reply:
I believe it works. We have tests showing that it works for:
- strings
- integers
- longs
- objects
- arrays
- maps
Do you think that it is necessery to cover all the types? If you go into imeplementation details at some point it is called ConvertToStructuredOrUnstructuredValue() method. And all the types that you mentioned are covered in some other tests - the tests for objects. So all these types are tested in context of objects.